graph/graph.go

```diff
@@ -485,10 +516,54 @@ func (graph *Graph) RawJSON(id string) ([]byte, error) {
 }
 
 func jsonPath(root string) string {
-	return filepath.Join(root, "json")
+	return filepath.Join(root, jsonFileName)
 }
 
 // TarLayer returns a tar archive of the image's filesystem layer.
 func (graph *Graph) TarLayer(img *image.Image) (arch archive.Archive, err error) {
-	return graph.driver.Diff(img.ID, img.Parent)
+	root := graph.imageRoot(img.ID)
+	mFileName := filepath.Join(root, tarDataFileName)
+	mf, err := os.Open(mFileName)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			logrus.Errorf("failed to open %q: %s", mFileName, err)
+		}
+		logrus.Debugf("[graph] TarLayer with traditional differ: %s", img.ID)
+		return graph.driver.Diff(img.ID, img.Parent)
+	}
+	pR, pW := io.Pipe()
+	// this will need to be in a goroutine, as we are returning the stream of a
+	// tar archive, but can not close the metadata reader early (when this
+	// function returns)...
+	go func() {
+		defer mf.Close()
+		// let's reassemble!
+		logrus.Debugf("[graph] TarLayer with reassembly: %s", img.ID)
+		mfz, err := gzip.NewReader(mf)
+		if err != nil {
+			pW.CloseWithError(fmt.Errorf("[graph] error with %s:  %s", mFileName, err))
+			return
```

Comment:
<3

On Tue, Jul 7, 2015 at 3:32 PM, Jonathan Boulle <notifications@github.com>
wrote:

> In graph/graph.go
> <https://github.com/docker/docker/pull/14067#discussion_r34080424>:
>
> > +		}
> > +		logrus.Debugf("[graph] TarLayer with traditional differ: %s", img.ID)
> > +		return graph.driver.Diff(img.ID, img.Parent)
> > +	}
> > +	pR, pW := io.Pipe()
> > +	// this will need to be in a goroutine, as we are returning the stream of a
> > +	// tar archive, but can not close the metadata reader early (when this
> > +	// function returns)...
> > +	go func() {
> > +		defer mf.Close()
> > +		// let's reassemble!
> > +		logrus.Debugf("[graph] TarLayer with reassembly: %s", img.ID)
> > +		mfz, err := gzip.NewReader(mf)
> > +		if err != nil {
> > +			pW.CloseWithError(fmt.Errorf("[graph] error with %s:  %s", mFileName, err))
> > +			return
>
> thanks :P
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/14067/files#r34080424>.
>